src/observe.ts

```diff
@@ -0,0 +1,330 @@
+			this._setObserver();
+		}
+
+		/**
+		* Initializes observation on the underlying object, preventing multiple changes to the same
+		* property from emitting multiple notifications.
+		*/
+		protected _setObserver(): void {
+			const store = this._propertyStore;
+			const target = this._target;
+
+			this._observeHandler = function (changes: ChangeEvent[]): void {
+				const propertyMap: Hash<number> = {};
+				const events: PropertyEvent[] = changes.reduce(function (
+					events: PropertyEvent[],
+					change: ChangeEvent
+				): PropertyEvent[] {
+					const property: string = change.name;
+
+					if (!this.onlyReportObserved || (property in store)) {
+						if (property in propertyMap) {
+							events.splice(propertyMap[property], 1);
+						}
+
+						propertyMap[property] = events.length;
+
+						events.push({
+							target: target,
+							name: property
+						});
+					}
+
+					return events;
+				}.bind(this), []);
+
+				if (events.length) {
+					this._listener(events);
+				}
+			}.bind(this);
+
+			(<any> Object).observe(target, this._observeHandler);
+		}
+
+		/**
+		* Ends all notifications on the target.
+		*/
+		destroy(): void {
+			const target = this._target;
+
+			(<any> Object).unobserve(target, this._observeHandler);
+			this._listener = this._observeHandler = this._propertyStore = this._target = null;
+		}
+
+		/**
+		* Enables notifications for the given property (or properties).
+		*
+		* If the `onlyReportObserved` option is `false`, then adding new properties will have no effect until
+		* `onlyReportObserved` is reset to `true`.
+		*
+		* @param properties The property name or arguments list of property names that will be observed.
+		*/
+		observeProperty(...properties: string[]): void {
+			const store = <any> this._propertyStore;
+
+			properties.forEach(function (property: string): void {
+				store[property] = 1;
+			});
+		}
+
+		/**
+		* Disables notifications for the given property (or properties).
+		*
+		* If the `onlyReportObserved` option is `false`, then removing properties will have no effect until
+		* `onlyReportObserved` is reset to `true`.
+		*
+		* * @param properties The property name or arguments list of property names that will be removed.
+		*/
+		removeProperty(...properties: string[]): void {
+			const store = this._propertyStore;
+
+			properties.forEach(function (property: string): void {
+				// Since the store is just a simple map, using the `delete` operator is not problematic.
+				delete (<any> store)[property];
+			});
+		}
+	}
+}
+
+export default function observe(kwArgs: ObserveArgs): Observer {
+	const Ctor = kwArgs.nextTurn && has('object-observe') ? ES7.Observer : ES5.Observer;
+
+	return new Ctor(kwArgs);
+}
+
+export interface ObserveArgs {
+	listener: (events: PropertyEvent[]) => any;
+	nextTurn?: boolean;
+	onlyReportObserved?: boolean;
+	target: {}
+}
+
+export interface Observer extends Handle {
```

Comment:
I was trying to preserve alphabetical order, which means `ES5` and `ES7` need to move before `KwArgs`...